In [2]:
import gc
gc.collect()

20

# Analysis of Android and iOS Applications

This project is an analysis of the popularity of the apps that are available on the google play store and Apple store for android and iOS devices respectively. The scope of the project is limited to the applications that are free and have english names

In [3]:
import pandas as pd
import numpy as np

In [4]:
# Reading the data
gg = pd.read_csv('D:/PERSONAL/DATASETS/googleplaystore.csv')


In [5]:
ios = pd.read_csv('D:/PERSONAL/DATASETS/AppleStore.csv')

Now that the dataset is loaded, let's have a look at the data.

In [6]:
gg.head()

App        Category  Rating  \
0     Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN     4.1   
1                                Coloring book moana  ART_AND_DESIGN     3.9   
2  U Launcher Lite – FREE Live Cool Themes, Hide ...  ART_AND_DESIGN     4.7   
3                              Sketch - Draw & Paint  ART_AND_DESIGN     4.5   
4              Pixel Draw - Number Art Coloring Book  ART_AND_DESIGN     4.3   

  Reviews  Size     Installs  Type Price Content Rating  \
0     159   19M      10,000+  Free     0       Everyone   
1     967   14M     500,000+  Free     0       Everyone   
2   87510  8.7M   5,000,000+  Free     0       Everyone   
3  215644   25M  50,000,000+  Free     0           Teen   
4     967  2.8M     100,000+  Free     0       Everyone   

                      Genres      Last Updated         Current Ver  \
0               Art & Design   January 7, 2018               1.0.0   
1  Art & Design;Pretend Play  January 15, 2018               2.0.0   
2               Art & Design    August 1, 2018               1.2.4   
3               Art & Design      June 8, 2018  Varies with device   
4    Art & Design;Creativity     June 20, 2018                 1.1   

    Android Ver  
0  4.0.3 and up  
1  4.0.3 and up  
2  4.0.3 and up  
3    4.2 and up  
4    4.4 and up

In [7]:
ios.head()

id               track_name  size_bytes currency  price  \
0  284882215                 Facebook   389879808      USD    0.0   
1  389801252                Instagram   113954816      USD    0.0   
2  529479190           Clash of Clans   116476928      USD    0.0   
3  420009108               Temple Run    65921024      USD    0.0   
4  284035177  Pandora - Music & Radio   130242560      USD    0.0   

   rating_count_tot  rating_count_ver  user_rating  user_rating_ver      ver  \
0           2974676               212          3.5              3.5     95.0   
1           2161558              1289          4.5              4.0    10.23   
2           2130805               579          4.5              4.5  9.24.12   
3           1724546              3842          4.5              4.0    1.6.2   
4           1126879              3594          4.0              4.5    8.4.1   

  cont_rating        prime_genre  sup_devices.num  ipadSc_urls.num  lang.num  \
0          4+  Social Networking               37                1        29   
1         12+      Photo & Video               37                0        29   
2          9+              Games               38                5        18   
3          9+              Games               40                5         1   
4         12+              Music               37                4         1   

   vpp_lic  
0        1  
1        1  
2        1  
3        1  
4        1

Features (columns) of each dataset:

In [8]:
gg.columns

Index(['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type',
       'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver',
       'Android Ver'],
      dtype='object')

In [9]:
ios.columns

Index(['id', 'track_name', 'size_bytes', 'currency', 'price',
       'rating_count_tot', 'rating_count_ver', 'user_rating',
       'user_rating_ver', 'ver', 'cont_rating', 'prime_genre',
       'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic'],
      dtype='object')

**Columns with potentially valuable insights in the google play store dataset:**
- Category: The genre of the app
- Rating: Rating of the app
- Installs: Number of Installations by the users

- Genres: Sub-genre of the app
- Content Rating: This rating is given based on the users for which the app would be appropriate to use.

**Columns with potentially valuable insights in the Apple Store dataset:**
- id: App identifier
- user_rating: Average rating given by the user.
- prime_genre: Primary genre of the app
- cont_rating: Age above which the application would be appropriate to be used by a user.
- sup_devices.num: Number of devices which support the app.

For more information regarding the features in the datasets please refer to the following links:

[ios AppleStore Dataset](https://www.kaggle.com/ramamet4/app-store-apple-data-set-10k-apps)  
[Google Play Store Dataset](https://www.kaggle.com/lava18/google-play-store-apps)

# Data Cleaning

**google playstore dataset**

In [10]:
gg.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10841 entries, 0 to 10840
Data columns (total 13 columns):
App               10841 non-null object
Category          10841 non-null object
Rating            9367 non-null float64
Reviews           10841 non-null object
Size              10841 non-null object
Installs          10841 non-null object
Type              10840 non-null object
Price             10841 non-null object
Content Rating    10840 non-null object
Genres            10841 non-null object
Last Updated      10841 non-null object
Current Ver       10833 non-null object
Android Ver       10838 non-null object
dtypes: float64(1), object(12)
memory usage: 1.1+ MB


Here, we can see that the  google play store dataset or the gg dataset has 10841 rows and 13 columns.  
We can also see that there are some columns which have null values as the the number of non-null values in such columns is not equal to 10840. These columns are:

- Rating
- Type
- Current Ver
- Android Ver

Although these columns have missing information, these rows of data will be handled as we proceed with the data cleaning process.

In [11]:
gg[gg.isna().any(axis=1)]
# Here we can see that there are 1481 missing values in total.

App             Category  Rating Reviews  \
15     Learn To Draw Kawaii Characters       ART_AND_DESIGN     3.2      55   
23              Mcqueen Coloring pages       ART_AND_DESIGN     NaN      61   
113          Wrinkles and rejuvenation               BEAUTY     NaN     182   
123             Manicure - nail design               BEAUTY     NaN     119   
126       Skin Care and Natural Beauty               BEAUTY     NaN     654   
...                                ...                  ...     ...     ...   
10824                        Cardio-FR              MEDICAL     NaN      67   
10825               Naruto & Boruto FR               SOCIAL     NaN       7   
10831         payermonstationnement.fr  MAPS_AND_NAVIGATION     NaN      38   
10835                         FR Forms             BUSINESS     NaN       0   
10838           Parkinson Exercices FR              MEDICAL     NaN       3   

       Size  Installs  Type Price Content Rating  \
15     2.7M    5,000+  Free     0       Everyone   
23     7.0M  100,000+  Free     0       Everyone   
113    5.7M  100,000+  Free     0   Everyone 10+   
123    3.7M   50,000+  Free     0       Everyone   
126    7.4M  100,000+  Free     0           Teen   
...     ...       ...   ...   ...            ...   
10824   82M   10,000+  Free     0       Everyone   
10825  7.7M      100+  Free     0           Teen   
10831  9.8M    5,000+  Free     0       Everyone   
10835  9.6M       10+  Free     0       Everyone   
10838  9.5M    1,000+  Free     0       Everyone   

                                Genres        Last Updated Current Ver  \
15                        Art & Design        June 6, 2018         NaN   
23     Art & Design;Action & Adventure       March 7, 2018       1.0.0   
113                             Beauty  September 20, 2017         8.0   
123                             Beauty       July 23, 2018         1.3   
126                             Beauty       July 17, 2018        1.15   
...                                ...                 ...         ...   
10824                          Medical       July 31, 2018       2.2.2   
10825                           Social    February 2, 2018         1.0   
10831                Maps & Navigation       June 13, 2018   2.0.148.0   
10835                         Business  September 29, 2016       1.1.5   
10838                          Medical    January 20, 2017         1.0   

      Android Ver  
15     4.2 and up  
23     4.1 and up  
113    3.0 and up  
123    4.1 and up  
126    4.1 and up  
...           ...  
10824  4.4 and up  
10825  4.0 and up  
10831  4.0 and up  
10835  4.0 and up  
10838  2.2 and up  

[1481 rows x 13 columns]

Let's first check if there is duplicate data regarding the same application.

In [12]:
# Getting the values of those apps whose count is more than 1 
duplicate = gg['App'].value_counts()[gg['App'].value_counts()>1]

In [13]:

len(duplicate)

798

There are total 798 rows which have duplicate data regarding the apps. However, it is necessary to check if the data in these rows are completely duplicated or if they are only partially duplicated

In [14]:
# Getting the names of apps for which duplicate data is present.
duplicate_apps = list(duplicate.index)

In [15]:
# Making a subset dataframe which has only the duplicate information
dups = gg[gg['App'].isin(duplicate_apps)]
dups.shape

(1979, 13)

Hence, there are 1979 rows of duplicate data

In [16]:
# Names of duplicate apps
set(dups['App'])

{'10 Best Foods for You',
 '1800 Contacts - Lens Store',
 '2017 EMRA Antibiotic Guide',
 '21-Day Meditation Experience',
 '365Scores - Live Scores',
 '420 BZ Budeze Delivery',
 '8 Ball Pool',
 '8fit Workouts & Meal Planner',
 '95Live -SG#1 Live Streaming App',
 'A Manual of Acupuncture',
 'A&E - Watch Full Episodes of TV Shows',
 'AAFP',
 'ABC News - US & World News',
 'AC - Tips & News for Android™',
 'AP Mobile - Breaking News',
 'ASCCP Mobile',
 'ASOS',
 'Accounting App - Zoho Books',
 'AccuWeather: Daily Forecast & Live Weather Reports',
 'Acorns - Invest Spare Change',
 'AdWords Express',
 'Ada - Your Health Guide',
 'Adobe Acrobat Reader',
 'Adobe Photoshop Express:Photo Editor Collage Maker',
 'Adult Dirty Emojis',
 'Advanced Comprehension Therapy',
 'Agar.io',
 'Airbnb',
 'Airway Ex - Intubate. Anesthetize. Train.',
 'AliExpress - Smarter Shopping, Better Living',
 'All Football - Latest News & Videos',
 'All Mental disorders',
 'All Social Networks',
 "Alto's Adventure",
 'Ama

In [17]:

dups[dups['App']=='Agar.io']

App Category  Rating  Reviews Size      Installs  Type Price  \
1947  Agar.io     GAME     4.2  3816799  32M  100,000,000+  Free     0   
3880  Agar.io     GAME     4.2  3815614  32M  100,000,000+  Free     0   

     Content Rating  Genres   Last Updated Current Ver   Android Ver  
1947       Everyone  Action  July 23, 2018       2.2.8  4.0.3 and up  
3880       Everyone  Action  July 23, 2018       2.2.8  4.0.3 and up

As we can see, there is quite a lot of similarity in the two rows except in the reviews column.  
Row number *1947* has more number of reviews than row number *3880*. This could mean that the data in row number *1947* is more recent.  
We can use this information as a criteria for removing the duplicate data so that the most recent data can be retained.

In [18]:
# Making a copy of the original dataset so that any error in the copy dataset will not affect the original dataset
gg1 = gg.copy()

The following loop is used to remove the duplicate rows that have reviews less than the highest number of reviews given to a particular app.

In [19]:
# Dropping the duplicate rows of an app whose count of reviews is less than the max number of reviews given to the app.
for i in range(len(duplicate_apps)):
    dup = duplicate_apps[i]
    max1 = gg1[gg1['App']==dup]
    max2 = max1['Reviews'].max()
    gg1.drop(gg1[(gg1['App']==dup)&(gg1['Reviews']<max2)].index,inplace=True)

In [20]:
gg1.shape

(10056, 13)

As mentioned above, there are a total of 1979 duplicate rows. However, one row for each app, which has the highest number of reviews must be retained to reduce the loss of data. Therefore, the actual reduction in the number of rows would be: *1979-798 = 1181*. Hence, the number of rows that are left are: *10841-1181 = 9660*

In [21]:
gg1.shape

(10056, 13)

As we can see that the number of rows in the dataset is still more than the required number of rows. This could mean that there are still duplicate data in the dataset

In [22]:
dup2 = gg1['App'].value_counts()[gg1['App'].value_counts()>1]

In [23]:
duplicate_apps2 = list(dup2.index)

In [24]:
len(duplicate_apps2)

337

In [25]:
gg1[gg1['App'].isin(duplicate_apps2)]

App             Category  \
164                                       Ebook Reader  BOOKS_AND_REFERENCE   
192                      Docs To Go™ Free Office Suite             BUSINESS   
193                                 Google My Business             BUSINESS   
204                                                Box             BUSINESS   
213                                ZOOM Cloud Meetings             BUSINESS   
...                                                ...                  ...   
8654   TickTick: To Do List with Reminder, Day Planner         PRODUCTIVITY   
8658                           ColorNote Notepad Notes         PRODUCTIVITY   
10049        Airway Ex - Intubate. Anesthetize. Train.              MEDICAL   
10473                          osmino Wi-Fi: free WiFi                TOOLS   
10768                                             AAFP              MEDICAL   

       Rating  Reviews                Size      Installs  Type Price  \
164       4.1    85842                 37M    5,000,000+  Free     0   
192       4.1   217730  Varies with device   50,000,000+  Free     0   
193       4.4    70991  Varies with device    5,000,000+  Free     0   
204       4.2   159872  Varies with device   10,000,000+  Free     0   
213       4.4    31614                 37M   10,000,000+  Free     0   
...       ...      ...                 ...           ...   ...   ...   
8654      4.6    25370  Varies with device    1,000,000+  Free     0   
8658      4.6  2401017  Varies with device  100,000,000+  Free     0   
10049     4.3      123                 86M       10,000+  Free     0   
10473     4.2   134203                4.1M   10,000,000+  Free     0   
10768     3.8       63                 24M       10,000+  Free     0   

      Content Rating             Genres    Last Updated         Current Ver  \
164         Everyone  Books & Reference   June 25, 2018               5.0.6   
192         Everyone           Business   April 2, 2018  Varies with device   
193         Everyone           Business   July 24, 2018    2.19.0.204537701   
204         Everyone           Business   July 31, 2018  Varies with device   
213         Everyone           Business   July 20, 2018      4.1.28165.0716   
...              ...                ...             ...                 ...   
8654        Everyone       Productivity  August 6, 2018  Varies with device   
8658        Everyone       Productivity   June 27, 2018  Varies with device   
10049       Everyone            Medical    June 1, 2018              0.6.88   
10473       Everyone              Tools  August 7, 2018             6.06.14   
10768       Everyone            Medical   June 22, 2018               2.3.1   

              Android Ver  
164            4.0 and up  
192    Varies with device  
193            4.4 and up  
204    Varies with device  
213            4.0 and up  
...                   ...  
8654   Varies with device  
8658   Varies with device  
10049          5.0 and up  
10473          4.4 and up  
10768          5.0 and up  

[733 rows x 13 columns]

As suspected, there are 733 duplicate rows of 337 apps in the dataset.

In [26]:
len(dup2)

337

In [27]:
dup2

Google Keep                                           4
WatchESPN                                             4
Skyscanner                                            4
Quizlet: Learn Languages & Vocab with Flashcards      4
eBay: Buy & Sell this Summer - Discover Deals Now!    4
                                                     ..
Fuzzy Numbers: Pre-K Number Foundation                2
Udacity - Lifelong Learning                           2
MailChimp - Email, Marketing Automation               2
Honkai Impact 3rd                                     2
LINE: Free Calls & Messages                           2
Name: App, Length: 337, dtype: int64

In [28]:
# checking the rows of one of the apps for which duplicate data is present.
gg1[gg1['App']=='WatchESPN']

App Category  Rating Reviews  Size     Installs  Type Price  \
3016  WatchESPN   SPORTS     4.1  288809  6.6M  10,000,000+  Free     0   
3022  WatchESPN   SPORTS     4.1  288809  6.6M  10,000,000+  Free     0   
3067  WatchESPN   SPORTS     4.1  288809  6.6M  10,000,000+  Free     0   
3094  WatchESPN   SPORTS     4.1  288809  6.6M  10,000,000+  Free     0   

     Content Rating  Genres        Last Updated Current Ver Android Ver  
3016       Everyone  Sports  September 27, 2017       2.5.1  4.4 and up  
3022       Everyone  Sports  September 27, 2017       2.5.1  4.4 and up  
3067       Everyone  Sports  September 27, 2017       2.5.1  4.4 and up  
3094       Everyone  Sports  September 27, 2017       2.5.1  4.4 and up

Here, we can see that the rows are a complete duplicate of each other. therefore, we can retain any one row for each app and discard the rest.

In [29]:
# Dropping duplicate rows of the same app where the number of reviews are also the same
gg1.drop_duplicates(subset='App',keep='first',inplace=True)

In [30]:
gg1.shape

(9660, 13)

In [31]:
len(gg1['App'].unique())

9660

As the scope of the dataset is limited to apps that have english names, apps that have non-english names will be removed.  
To check the names of such apps, we will check the ASCII values of each character in name of an app.  
The **ASCII** values of the english alphabet ranges from *0 to 127*.  
Any app which has a character whose ASCII value does not lie in this range will be removed.  
It is possible that an app may have an english name but it may have characters whose ASCII values that do not fall in the above mentioned range. Therefore, all apps that have more than 3 characters whose ASCII values are not in this range will be removed.

In [32]:
# Counting the number of apps that have a non-english name in the gg dataset
count = 0
for i in range(gg1.shape[0]):
    name = gg1['App'].iloc[i]
    cnt = 0
    for j in name:
        if((ord(j)>127) or (ord(j)<0)):
            cnt+=1
            if(cnt>3):
                print(name)
                count+=1
                break

        else:
            continue
            
print('count:',count)

Flame - درب عقلك يوميا
သိင်္ Astrology - Min Thein Kha BayDin
РИА Новости
صور حرف H
L.POINT - 엘포인트 [ 포인트, 멤버십, 적립, 사용, 모바일 카드, 쿠폰, 롯데]
RMEduS - 음성인식을 활용한 R 프로그래밍 실습 시스템
AJ렌터카 법인 카셰어링
Al Quran Free - القرآن (Islam)
中国語 AQリスニング
日本AV历史
Ay Yıldız Duvar Kağıtları
বাংলা টিভি প্রো BD Bangla TV
Cъновник BG
CSCS BG (в български)
뽕티비 - 개인방송, 인터넷방송, BJ방송
BL 女性向け恋愛ゲーム◆俺プリクロス
SecondSecret ‐「恋を読む」BLノベルゲーム‐
BL 女性向け恋愛ゲーム◆ごくメン
あなカレ【BL】無料ゲーム
감성학원 BL 첫사랑
BQ-መጽሐፍ ቅዱሳዊ ጥያቄዎች
BS Calendar / Patro / पात्रो
Vip视频免费看-BT磁力搜索
Билеты ПДД CD 2019 PRO
Offline Jízdní řády CG Transit
Bonjour 2017 Abidjan CI ❤❤❤❤❤
CK 初一 十五
الفاتحون Conquerors
DG ग्राम / Digital Gram Panchayat
DM הפקות
DW فارسی By dw-arab.com
لعبة تقدر تربح DZ
বাংলাflix
RPG ブレイジング ソウルズ アクセレイト
英漢字典 EC Dictionary
ECナビ×シュフー
أحداث وحقائق | خبر عاجل في اخبار العالم
EG SIM CARD (EGSIMCARD, 이지심카드)
パーリーゲイツ公式通販｜EJ STYLE（イージェイスタイル）
FAHREDDİN er-RÂZİ TEFSİRİ
I'm Rich/Eu sou Rico/أنا غني/我很有錢
AÖF Ev İdaresi 1. Sınıf
Ey Sey Storytime រឿងនិទានតាឥសី
哈哈姆特不EY
FP Разбиты

There are 45 apps that have non-english names. These apps will be removed from the dataset.

In [33]:
# Dropping the rows that have apps that have a non-english name in the gg dataset
count = 0
i = gg1.shape[0]-1
while i > 0:
    cnt = 0
    name = gg1['App'].iloc[i]
    
    for j in name:
        if((ord(j)>127) or (ord(j)<0)):
            cnt+=1
            if(cnt>3):
                gg1.drop((gg1[gg1['App']==name]).index,inplace=True)
                gg1.reset_index(drop=True,inplace=True)
                count+=1
                break

    i-=1
            
print('count:',count)

count: 45


In [34]:
# The number of rows that should be left after removal of these apps will be:
9660-45

9615

In [35]:
gg1.shape

(9615, 13)

After removing the duplicate data, the apps which are not free of cost will be removed.

In [36]:
# Removing the apps which are not free.

gg1.columns

Index(['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type',
       'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver',
       'Android Ver'],
      dtype='object')

In [37]:
(gg1['Price'].dtype)

dtype('O')

In [38]:
price = gg1['Price'].copy()

In [39]:
# Removing the dollar sign so that the column can be coverted to a numeric data type.
price = price.str.lstrip('$')

In [40]:
price[price!='0']

218      4.99
219      4.99
329      3.99
362      6.99
363      1.49
        ...  
9511     0.99
9536     7.99
9556    16.99
9559     1.20
9572     1.04
Name: Price, Length: 753, dtype: object

In [41]:
gg1['Price'] = price

In [42]:
gg1[gg1['Price']!='0']

App            Category  \
218   TurboScan: scan documents and receipts in PDF            BUSINESS   
219                  Tiny Scanner Pro: PDF Doc Scan            BUSINESS   
329                              Puffin Browser Pro       COMMUNICATION   
362                                      Calculator              DATING   
363                               Truth or Dare Pro              DATING   
...                                             ...                 ...   
9511                                    FP VoiceBot              FAMILY   
9536                                Fast Tract Diet  HEALTH_AND_FITNESS   
9556                        Trine 2: Complete Story                GAME   
9559                                   sugar, sugar              FAMILY   
9572                           Word Search Tab 1 FR              FAMILY   

      Rating Reviews                Size  Installs  Type  Price  \
218      4.7   11442                6.8M  100,000+  Paid   4.99   
219      4.8   10295                 39M  100,000+  Paid   4.99   
329      4.0   18247  Varies with device  100,000+  Paid   3.99   
362      2.6      57                6.2M    1,000+  Paid   6.99   
363      NaN       0                 20M       50+  Paid   1.49   
...      ...     ...                 ...       ...   ...    ...   
9511     NaN      17                157k      100+  Paid   0.99   
9536     4.4      35                2.4M    1,000+  Paid   7.99   
9556     3.8     252                 11M   10,000+  Paid  16.99   
9559     4.2    1405                9.5M   10,000+  Paid   1.20   
9572     NaN       0               1020k       50+  Paid   1.04   

     Content Rating            Genres       Last Updated  Current Ver  \
218        Everyone          Business     March 25, 2018        1.5.2   
219        Everyone          Business     April 11, 2017        3.4.6   
329        Everyone     Communication       July 5, 2018  7.5.3.20547   
362        Everyone            Dating   October 25, 2017        1.1.6   
363            Teen            Dating  September 1, 2017          1.0   
...             ...               ...                ...          ...   
9511     Mature 17+     Entertainment  November 25, 2015          1.2   
9536       Everyone  Health & Fitness     August 8, 2018        1.9.3   
9556           Teen            Action  February 27, 2015         2.22   
9559       Everyone            Puzzle       June 5, 2018          2.7   
9572       Everyone            Puzzle   February 6, 2012          1.1   

     Android Ver  
218   4.0 and up  
219   3.0 and up  
329   4.1 and up  
362   4.0 and up  
363   4.0 and up  
...          ...  
9511  2.1 and up  
9536  4.2 and up  
9556  5.0 and up  
9559  2.3 and up  
9572  3.0 and up  

[753 rows x 13 columns]

In [43]:
gg1.dtypes

App                object
Category           object
Rating            float64
Reviews            object
Size               object
Installs           object
Type               object
Price              object
Content Rating     object
Genres             object
Last Updated       object
Current Ver        object
Android Ver        object
dtype: object

In [44]:
gg1['Price'] = gg1['Price'].astype('float')

ValueError: could not convert string to float: 'Everyone'

In [ ]:
# The error in the above cell indicates that one of the rows has inappropriate type of data in the Price column.
gg1[gg1['Price']=='Everyone']

In [ ]:
# Dropping the row which has the value 'Everyone' in the price column.
gg1.drop((gg1[gg1['Price']=='Everyone']).index,axis=0,inplace=True)

In [ ]:
gg1['Price'] = gg1['Price'].astype('float')

In [ ]:
# Dropping the rows that have apps which are not free of cost.
gg1.drop((gg1[gg1['Price']>0]).index,axis=0,inplace=True)

In [ ]:
gg1.shape

In [ ]:
set(gg1['Type'])

In [ ]:
# Number of NA values in the dataset.
gg1.isna().sum()

There are still null values left after removing duplicate data and apps that are not free of cost.

In [ ]:
# Total number of na values.
gg1.isna().sum().sum()

In [ ]:
gg1.info()

In [ ]:
gg1[gg1['Type'].isna()]

As there is a large number of null values in the rating column, removing all such rows will result in a substantial loss of data. Therefore, instead of dropping the rows, the null value in the rating column will replaced with the average rating given to that category of apps. 

In [ ]:
# Taking a subset of the dataset in which the Rating column has only NA values.
x = gg1[gg1['Rating'].isna()]

In [ ]:
# Creating a list of categories which has rows in which the Rating column has NA values.
cat = list(set(x['Category']))

In [ ]:
for i in range(len(cat)):
    avg_rating = gg1['Rating'][gg1['Category']==cat[i]].mean()
    gg1.Rating[(gg1['Rating'].isna()) & (gg1['Category']==cat[i])]=avg_rating

In [ ]:
gg1.Rating[(gg1['Rating']==np.nan) & (gg1['Category']=='BEAUTY')]

In [ ]:
gg1.info()

In [ ]:
# Number of NA values left
gg1.isna().sum().sum()

In [ ]:
gg1[gg1.isna().any(axis=1)]

In [ ]:
set(gg1['Current Ver'])

After replacing the null values in the *Rating* column, there are 8 null values in the *Current Ver* column. As the version of each app is distinct and unique, it cannot be replaced by another value from the column. Therefore, these values will be dropped

In [ ]:
gg1.dropna(inplace=True)

In [ ]:
gg1.isna().sum().sum()

In [ ]:
gg1.info()

# Data Cleaning
**ios dataset**

Now, we will be performing the same data cleaning procedure for the ios dataset.

In [ ]:
ios.head()

Checking for rows that are duplicates:

In [ ]:
ios['track_name'].value_counts()[ios['track_name'].value_counts()>1]

In [ ]:
ios.shape

In [ ]:
ios[ios['track_name']=='Mannequin Challenge']

In [ ]:
ios[ios['track_name']=='VR Roller Coaster']

In this case, we will be retaining those rows which have the highest *rating_count_tot*. This column contains the number of ratings given by the users to a particular app for all versions. Here, an assumption is taken that, more the number of ratings given, more relevant and recent is the information.

In [ ]:
ios_dups = list(ios['track_name'].value_counts()[ios['track_name'].value_counts()>1].index)

Dropping the duplicate rows:

In [ ]:
for i in range(len(ios_dups)):
    sub = ios[ios['track_name']==ios_dups[i]]
    max2 = sub['ver'].max()
    ios.drop(ios[(ios['track_name']==ios_dups[i]) & (ios['ver']!=max2)].index,inplace=True)
    

In [ ]:
ios[ios['track_name']=='Mannequin Challenge']

In [ ]:
ios.shape

Now, let's count the number of apps that have non-english names with the same condition that was applied in the google play store dataset i.e. those apps which have more than 3 characters in their names whose ASCII values do not lie in the range of 0 to 127 will be removed.

In [ ]:
# Counting the number of apps that have a non-english name in the ios dataset
count = 0
for i in range(ios.shape[0]):
    cnt = 0
    name = ios['track_name'].iloc[i]
    
    for j in name:
        if((ord(j)>127) or (ord(j)<0)):
            cnt+=1
            if(cnt>3):
                print(name)
                count+=1
                break
            
            
        else:
            continue
            
print('count:',count)

In [ ]:
# Number of apps to be left in the ios dataset after removal of non-english apps
7195-1014

In [ ]:
ios.shape

In [ ]:
# Dropping the rows that have apps that have a non-english name in the ios dataset
count = 0
i = ios.shape[0]-1
while i > 0:
    cnt = 0
#     print('i:',i)
    name = ios['track_name'].iloc[i]
    
    for j in name:
        if((ord(j)>127) or (ord(j)<0)):
            cnt+=1
            if(cnt>3):
                print(name)
#                 print('dropping')
                ios.drop((ios[ios['track_name']==name]).index,inplace=True)
                ios.reset_index(drop=True,inplace=True)
    #             i = ios.shape[0]-1
#                 print('if i:',i)
#                 print('dropped')
                count+=1
                break

    i-=1
            
print('count:',count)

Next, the apps that are not free of cost will be removed.

In [ ]:
# Removing apps that are not free.
ios[ios['price']>0]

As there are 2961 apps that are not free, after their removal, the number of rows left will be: *6181-2961 = 3220*

In [ ]:
ios.drop((ios[ios['price']>0]).index,axis=0,inplace=True)

In [ ]:
print(ios.shape)
print(gg1.shape)

In [ ]:
ios.info()

In [ ]:
ios.isna().sum()

In [ ]:
gg1.isna().sum()

The data cleaning process of both the datasets is now complete. During this process, it is possible that certain variables were created but were not used. It is important to delete those variables in order to free the memory occupied by them for the smooth execution of our program.

In [ ]:
import gc
gc.collect()

Now we can begin the analysis of the datasets. First, we will check the most common category of apps in both the dataset

# Frequency Tables

A frequency table for each dataset will be created which has the count for each category of apps available in both the stores.

**gg1 dataset**

In [ ]:
gg1.columns

In [ ]:
import matplotlib.pyplot as plt
import seaborn as sns
import matplotlib as mpl

In [ ]:
plt.rcParams['font.size']=20

In [ ]:
fig,ax = plt.subplots(1,1,figsize=(100,30))
plt.hist(gg1['Category'],ec='black',alpha=0.5,rwidth=0.45,bins=33)
plt.yticks(fontsize=45)
plt.xticks(fontsize=25,rotation=15,ticks=list(range(0,34)))
plt.grid(axis='y',color='black',linewidth=2)




As we can see that the maximum applications are in the category of *FAMILY* which has over 1600 apps followed by *GAME* which has over 800 apps.

Let us see what are the popular sub-genres of these two categories.

In [ ]:
family_game = gg1[(gg1['Category']=='FAMILY') | (gg1['Category']=='GAME')]

In [ ]:
len(set(family_game['Genres']))

In [ ]:
fig,ax = plt.subplots(1,1,figsize=(100,30))

plt.hist(family_game['Genres'],bins=79,ec='black',alpha=0.5,rwidth=0.5)
plt.grid(axis='y',linewidth=2)
plt.xticks(rotation=45,ha='right',ticks=list(range(0,80)),rotation_mode='anchor')
plt.show()


The most popular sub-genres for the FAMILY and GAME categories are Entertainment and Education as both of them have over **400** apps and over **300** apps respectively. 

**ios dataset**

In [ ]:
ios.columns

In [ ]:
len(set(ios['prime_genre']))

In [ ]:
y_val = np.arange(0,2020,500)
fig,ax = plt.subplots(1,1,figsize=(20,5))
plt.hist(ios['prime_genre'],ec='black',alpha=0.5,rwidth=0.3,bins=23)
plt.yticks(y_val,fontsize=10)
plt.xticks(fontsize=10,rotation=-20)
plt.grid(axis='y',color='black',linewidth=1)




Here, the number of applications of the genre *Games* clearly stands out from the number of applications of other categories.

Although the above histograms do show some valuable insghts regarding the popularity of different types of in the two app stores, having more precise results regarding the applications would lead to a better understanding of the popularity of the different apps in these app stores.  

Keeping this in mind, let us now build a function which will give us the exact number of applications of each category of app as well the percentage of the apps of that category present in a particular store.

In [ ]:
def freq_table(df,col_name,percentage=False):
    dim = (df.shape[0])
    
    num_category = dict()
    for i in range(df.shape[0]):
        if(df[col_name].iloc[i] in num_category):
            num_category['{}'.format(df[col_name].iloc[i])]+=1
            
        else:
            num_category['{}'.format(df[col_name].iloc[i])] = 1
            
        
    keys = list(num_category.keys())
    values = list(num_category.values())
            
            
    if(percentage==False):      
        import pandas as pd

        labels = list(num_category.keys())
        values = list(num_category.values())
        num_category = pd.DataFrame({'Category':labels,'value':values})
        return num_category
            
    if(percentage==True):
        import pandas as pd

        num_labels = list(num_category.keys())
        num_values = list(num_category.values())
        num_category = pd.DataFrame({'Category':num_labels,'value':num_values})
        num_category.sort_values(ascending=False,by='value',inplace=True)
        num_category.reset_index(drop=True,inplace=True)
        percent_category = dict()
        
        for j in range(len(keys)):
            
            percent_category[keys[j]] = values[j]/dim          
            
        per_labels = list(percent_category.keys())
        per_values = list(percent_category.values())
        
        percent_category = pd.DataFrame({'Category':per_labels,'Percentage':per_values})
        percent_category.sort_values(ascending=False,by='Percentage',inplace=True)
        percent_category.reset_index(drop=True,inplace=True)
        return (num_category,percent_category)
    
    
        

In [ ]:
(gg_category,gg_percent_category) = freq_table(gg1,'Category',True)

In [ ]:
from IPython.display import display_html 
gg_category_style = gg_category.style.set_table_attributes("style='display:inline'").set_caption('Number of apps in each category')
gg_percentage_category_style = gg_percent_category.style.set_table_attributes("style='display:inline'").set_caption('Percentage of apps of each category')

In [ ]:
display_html(gg_category_style._repr_html_()+"\xa0\xa0\xa0\xa0"+gg_percentage_category_style._repr_html_(),raw=True)

Hence, we can see that there are **1675** apps of the category *FAMILY* which and nearly **19%** of all apps are of the category *FAMILY*. This is followed by the category *GAME* which is the second highest category in terms of the number of apps in each category after *FAMILY*. The number of apps which are of the *GAME* category are nearly half of that of the *FAMILY* category and consist approximately **10%** of all the apps in the google play store.

In [ ]:
(gg_genres,gg_percent_genres) = freq_table(family_game,'Genres',True)

In [ ]:
# from IPython.display import display_html
gg_genres_style = gg_genres.style.set_table_attributes("style='display:inline'").set_caption('Number of applications of each genre')
gg_percent_genres_style = gg_percent_genres.style.set_table_attributes("style='display:inline'").set_caption('Percentage of applications of each genre')

display_html(gg_genres_style._repr_html_()+'\xa0\xa0\xa0'+gg_percent_genres_style._repr_html_(),raw=True)

Here we can see that in the FAMILY and the GAME category, the most popular sub-category apps of these categories are apps of the *Entertainment* category followed by the *Education* category.  
There are a total of **2534** apps which are of the category FAMILY or GAME out of which **18%** of the apps are of the  Entertainment sub-category and approximately **15%** of the apps are of the Education sub-category.

In [ ]:
display_html(ios_genres_style._repr_html_()+"\xa0\xa0\xa0\xa0"+ios_percent_genres_style._repr_html_(),raw=True)

In [ ]:
sum(ios_genres['value'])

In the App store which contains the ios apps, the highes number of apps are of the category *Games* which consists of **1872 apps out 3220**. The apps of this category approximately **58%** of all the apps in the app store

In [ ]:
ios.head()

In [ ]:
ios.info()

Now, we will calculate the average number of installations that have been done for each genre in both the datasets.

In [ ]:
installs = gg1['Installs']

In [ ]:
gg1['Installs'] = installs.str.replace(',','')

In [ ]:
gg1.head()

In [ ]:
# Removing the special characters from the installs columns for its conversion to a numeric data type from a string data type.
gg1['Installs'] = gg1['Installs'].str.replace('+','')

In [ ]:
gg1.info()

In [ ]:
gg1.head()

In [ ]:
gg1['Installs'] = gg1['Installs'].astype('int')

In [ ]:
gg1.info()

The below loop will calculate the average rating for each category in the google play store dataset.

In [ ]:
avg_num_user = dict()

for i in range(gg_category.shape[0]):
    gen = gg_category['Category'].iloc[i]
    sub = gg1[gg1['Category']==gen]
    
    tot_installs = sum(sub['Installs'])
    
    num_of_apps = gg_category[gg_category['Category']== gg_category['Category'].iloc[i]]['value'].iloc[0]
    
    avg = int(tot_installs/num_of_apps)
    
    avg_num_user['{}'.format(gg_category['Category'].iloc[i])]=avg

In [ ]:
labels = list(avg_num_user.keys())
values = list(avg_num_user.values())
gg_avg_user = pd.DataFrame({'Category':labels,'Average number of installs':values})

In [ ]:
gg_avg_user.sort_values(ascending=False,inplace=True,by='Average number of installs')

In [ ]:
gg_avg_user

The above dataframe shows that event though the number of apps of the category *FAMILY* is the highest, it does not have the highest number of installs in the google play store. The apps of the category *COMMUNICATION* have the highest number of installations.

The following loop will calculate the average number of installs for each genre in the app store dataset.

In [ ]:
ios_avg_num_installs = dict()

for i in range(ios_genres.shape[0]):
    gen = ios_genres['Category'].iloc[i]
    
    sub = ios[ios['prime_genre']==gen]
    
    tot_installs = sum(sub['rating_count_tot'])
    
    num_of_apps = ios_genres[ios_genres['Category']== ios_genres['Category'].iloc[i]]['value'].iloc[0]
    
    avg = int(tot_installs/num_of_apps)
    
    ios_avg_num_installs['{}'.format(ios_genres['Category'].iloc[i])]=avg

In [ ]:
ios_avg_num_installs

In [ ]:
x = list(ios_avg_num_installs.keys())
y = list(ios_avg_num_installs.values())



ios_avg_installs = pd.DataFrame({'Genres':x,'Average Number of Installs':y})
ios_avg_installs.sort_values(ascending=False,by = 'Average Number of Installs',inplace=True)

In [ ]:
ios_avg_installs

In the app store, the number of apps which were highest were of the genre *Games*. However, the genre *NAVIGATION*  has the highest average number of installations.

Since the trend regarding the number of average installations in both the datasets is similar, it seems to indicate that the usage of an app of a particular genre does not depend on the number of apps available for that genre in the store.

Next, we will be analysing the apps on the basis of the average rating given to each genre in both the dataset.

The following loop will calculate the average rating given to each Category in the google play store dataset.

In [ ]:
avg_num_rating = dict()

for i in range(gg_category.shape[0]):
    gen = gg_category['Category'].iloc[i]
    sub = gg1[gg1['Category']==gen]
    
    tot_rating = sum(sub['Rating'])
    
    num_of_apps = gg_category[gg_category['Category']== gg_category['Category'].iloc[i]]['value'].iloc[0]
    
    avg = (tot_rating/num_of_apps)
    
    avg_num_rating['{}'.format(gg_category['Category'].iloc[i])]=avg

In [ ]:
labels = list(avg_num_rating.keys())
values = list(avg_num_rating.values())
gg_avg_rating = pd.DataFrame({'Category':labels,'Average Rating':values})

In [ ]:
gg_avg_rating.sort_values(by='Average Rating',ascending=False,inplace=True)

In [ ]:
gg_avg_rating

In [ ]:
ios.head()

The loop in the following cell will calculate the average rating of each genre in the ios dataset.

In [ ]:
ios_avg_num_rating = dict()

for i in range(ios_genres.shape[0]):
    gen = ios_genres['Category'].iloc[i]
    
    sub = ios[ios['prime_genre']==gen]
    
    tot_rating = sum(sub['user_rating'])
    
    num_of_apps = ios_genres[ios_genres['Category']== ios_genres['Category'].iloc[i]]['value'].iloc[0]
    
    avg = (tot_rating/num_of_apps)
    
    ios_avg_num_rating['{}'.format(ios_genres['Category'].iloc[i])]=avg

In [ ]:
x = list(ios_avg_num_rating.keys())
y = list(ios_avg_num_rating.values())



ios_avg_rating = pd.DataFrame({'Genres':x,'Average Rating':y})
ios_avg_rating.sort_values(ascending=False,by = 'Average Rating',inplace=True)

In [ ]:
ios_avg_installs

In both the datasets, lifestyle-based apps i.e. the *EVENTS* category in the google play store dataset and the *Catalog* genre in the app store dataset have been given the highest rating. This seems to indicate that although the number of installations for the *COMMUNICATION* category apps in the google play store dataset and the *Navigation* category apps in the app store dataset have been high but the users may not have been satisfied with the quality of the apps available in those categories, resulting in those not apps not receiving the rating which would be commensurate with the average number of installations observed in these apps.  
  
Another possible reason for the *EVENTS* and the *Catalog* category apps not receiving the highest average rating in their respective datasets is that the users who rated those apps may be very different from the users who rated the apps which have received the highest rating. Howver, this line of reasoning cannot be verified because of lack of appropriate data.